Fix tooltip descriptions for spectators.

matjaeck opened the pull request:
To fix #15200, regression of #15144.

MustaphaTR commented:
Well this definitely fixes the crash, but i'm not sure if this is a good idea. I would prefer specs to also see the tooltip descs in my mod. In any case i guess #15160 can fix that properly.

pchote commented:
It would be better if this can follow the same pattern we do in most other cases: use the RenderPlayer unless that is null, falling back to Localplayer, falling back to assuming that the stance is valid.

lawando commented:
Other than I expected the tooltip description can be shown to spectators if set to true. Please pardon my lack of understanding and knowledge, it's a learning process for me and I'm willing to improve. 

It will be required to control which tooltip descriptions in case of multiple instances of the trait should be shown to spectators or add a description that is only shown to spectators. My current idea is to add a "generic description" that is not aimed at players of a certain stance, integrate `ITooltipInfo` and use `TooltipForPlayerStance` to return the generic description for `stance == Stance.None`. If I understood it correctly then this is similar to how `Tooltip` filters out the generic prefix for spectators.

pchote commented:
https://github.com/OpenRA/OpenRA/blob/5b47aa4a9ef5baea90d3f64570c4bacefd69adf1/OpenRA.Mods.Common/Widgets/Logic/Ingame/WorldTooltipLogic.cs#L68 is how we do this for the generic stance prefixes.  Using the same thing here will ensure consistent behaviour.

lawando commented:
Still WIP, haven't found out yet how to filter the descriptions for spectators.

GraionDilach commented:
That's expected, see #10892.

lawando commented:
It looks like a proper implementation requires the spectator stances from #15160 which are assigned to next +1 while this is assigned to next release. IMO showing all descriptions to spectators looks broken so I suggest we either disable them for spectators until we have the spectator stances (preferably) or we revert #15144 entirely.

GraionDilach commented on OpenRA.Mods.Common/Traits/TooltipDescription.cs:
```diff
@@ -41,7 +49,7 @@ public bool IsTooltipVisible(Player forPlayer)
 				return false;
 
-			var stance = forPlayer.Stances[self.Owner];
+			var stance = Owner == null || forPlayer == null ? Stance.None : Owner.Stances[forPlayer];
```
You coulda try hardcoding this to Stance.Allies instead of Stance.None for now and adding a comment with a HACK prefix above it to mention it being a workaround. It might hide the neutral/enemy tooltips - I haven't tested if that would work.

GraionDilach replied:
If that doesn't work, then just hide them all for spectators IMO.

lawando commented:
I would appreciate if somebody could take over this PR. It's not that I lost interest in it but I doubt that I'll be wble to work on it in the next days as I want to focus on OpenRA/OpenRAWeb/381.

GraionDilach commented:
I'll do it on Sunday at worst.

pchote commented:
@reaperrr: Would you be able to take this one over?

matjaeck commented:
The other project is in a state where I have to wait for a review so I'll give this another try in the meantime.

matjaeck commented:
Updated, but still working on it.

matjaeck commented:
Will leave it like this and wait for a review. 

![101](https://user-images.githubusercontent.com/37185497/42628198-a119d8d2-85cf-11e8-9167-0685d6cd607f.gif)

pchote commented:
The default spectator view now shows *all* the tooltip options, in a rather broken way:
![screen shot 2018-07-25 at 22 29 45](https://user-images.githubusercontent.com/167819/43228784-6a07e98e-905a-11e8-9696-7ab16c719eb6.png)

pchote commented:
I suspect the way to fix this will be to remove `Stance.None` from the defaults, and then list that explicitly in the yaml definition for the "provides additional funds" label.

pchote commented:
Or alternately remove `Stance.None` from the defaults, and leave them disabled for spectators in the default view.  This could then be revisited after @GraionDilach's player stance overhaul.

GraionDilach commented:
I will reopen that PR after we get through the playtest.

matjaeck commented:
Updated. Removed `Stance.None` from `ValidStances ` again, however I can not reproduce the issue shown in the picture above. 

![101](https://user-images.githubusercontent.com/37185497/43272932-7524dac4-90fb-11e8-9e0a-550ee15d73e9.gif)

Spectators will see the descriptions for either the neutral stance or the allied stance. The reason that I removed `Stance.None` from the valid stances is that if you write a yaml definition for `TooltipDescription : ValidStances: None `, this description will not be shown to spectators.

abcdefg30 commented:
> OpenRA.Mods.Common/Traits/TooltipDescription.cs:L51: [SP2000] Invalid spacing at the end of the line.

matjaeck commented:
Fixed the style error.

Smittytron reviewed (dismissed):
Tested, LGTM

GraionDilach commented on OpenRA.Mods.Common/Traits/TooltipDescription.cs:
```diff
@@ -21,7 +21,7 @@ public class TooltipDescriptionInfo : ConditionalTraitInfo
 
 		[Desc("Player stances who can view the description.")]
-		public readonly Stance ValidStances = Stance.Ally | Stance.Neutral | Stance.Enemy;
+		public readonly Stance ValidStances = Stance.Enemy | Stance.Neutral | Stance.Ally;
```
This ordering is bad. :S

matjaeck replied:
The previous order caused my IDE to show an error so I changed it to the order used in the Enum, see https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/Traits/TraitsInterfaces.cs#L52.

matjaeck replied:
Fixed the stance order.

pchote commented:
The issue I posted above still happens for the "Disable Shroud" view.  This happens because `Stance.None` has a value of 0, and `N & 0 == 0` is true for any value of N.

Commit:
Fix tooltip descriptions for spectators.

pchote reviewed (dismissed):
I have force pushed my suggested workaround. :+1: with that now in place.

GraionDilach commented:
:+1: after Travis confirms the build works.

abcdefg30 commented:
> OpenRA.Mods.Common/Traits/TooltipDescription.cs:L55: [SP2000] Invalid spacing at the end of the line.

GraionDilach commented:
👍 after Travis confirms the build works. 😄

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/4bb35886589323348cdb613b04b10c72f0d7fc99...2b428ef719e4ac1fa8a10adeff9bd9a55f784860)
